components/authz-service/storage/v2/postgres/postgres.go

```diff
@@ -1274,6 +1269,94 @@ func (p *pg) ListRulesForProject(ctx context.Context, projectID string) ([]*v2.R
 	return rules, nil
 }
 
+func (p *pg) ApplyStagedRules(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	tx, err := p.db.BeginTx(ctx, nil)
+	if err != nil {
+		return p.processError(err)
+	}
+
+	// Upsert all staged rules into applied rules marked for update, returning the id and db_id
+	// of all rules affected so we can update their conditions below.
+	rows, err := tx.QueryContext(ctx,
+		`INSERT INTO iam_project_rules (id, project_id, name, type)
+				SELECT s.id, s.project_id, s.name, s.type
+					FROM iam_staged_project_rules AS s
+					WHERE deleted=false
+				ON CONFLICT (id) DO UPDATE
+				SET name=excluded.name, type=excluded.type
+				RETURNING id, db_id;`)
+	defer func() {
+		if err := rows.Close(); err != nil {
+			p.logger.Warnf("failed to close db rows: %s", err.Error())
+		}
+	}()
+	if err != nil {
+		return p.processError(err)
+	}
+
+	// For every staged rule updated, we need to update conditions.
+	ids := make(map[string]string)
+	for rows.Next() {
+		var id string
+		var dbID string
+		err = rows.Scan(&id, &dbID)
+		if err != nil {
+			return p.processError(err)
+		}
+		ids[id] = dbID
+	}
+
+	for id, dbID := range ids {
+		_, err = tx.ExecContext(ctx,
+			`DELETE FROM iam_rule_conditions WHERE rule_db_id=$1;`, dbID)
+		if err != nil {
+			return p.processError(err)
+		}
+
+		_, err = tx.ExecContext(ctx,
+			`INSERT INTO iam_rule_conditions (rule_db_id, value, attribute, operator)
+					SELECT $2, cond.value, cond.attribute, cond.operator FROM iam_staged_project_rules AS r
+					LEFT OUTER JOIN iam_staged_rule_conditions AS cond
+					ON rule_db_id=r.db_id
+					WHERE r.id=$1;`,
+			id, dbID,
+		)
+		if err != nil {
+			return p.processError(err)
+		}
+	}
+
```

Comment:
❓ Shouldn't something else happen here, too? Like, do the re-tagging stuff in the domain services? Not saying that this should belong in this PR, just wondering how we'd architect that.

⚠️ There's _nothing_ ensuring that this GRPC method isn't called twice concurrently; the transaction will only ensure that one of the `n` calls succeeds -- so, if we do something to some other service, that case needs to be considered.